test/BilateralSwap/bilateralSwap.js

```diff
@@ -0,0 +1,152 @@
+        });
+
+        it('Validate failure for incorrect input note values (k1 != k3, k2 != k4)', async () => {
+            const makerNoteValues = [10, 50];
+            const takerNoteValues = [20, 20];
+            const incorrectTestNoteValues = helpers.makeTestNotes(makerNoteValues, takerNoteValues);
+            
+            const { proofData, challenge } = bilateralProof.constructBilateralSwap(incorrectTestNoteValues, accounts[0]);
+
+            await exceptions.catchRevert(BilateralSwap.validateBilateralSwap(proofData, challenge, t2, {
+                from: accounts[0],
+                gas: 4000000,
+            }));
+        });
+
+        it('Validate failure for incorrect number of input notes', async () => {
+            const makerNoteValues = [10, 20, 30];
+            const takerNoteValues = [10, 20, 30];
+            const incorrectNumberOfNotes = helpers.makeTestNotes(makerNoteValues, takerNoteValues);
+            
+            const { proofData, challenge } = bilateralProof.constructBilateralSwap(incorrectNumberOfNotes, accounts[0]);
+
+            await exceptions.catchRevert(BilateralSwap.validateBilateralSwap(proofData, challenge, t2, {
+                from: accounts[0],
+                gas: 4000000,
+            }));
+        });
+
+        it('Validate failure for a bid note of zero value', async () => {
+            const makerNoteValues = [0, 20];
+            const takerNoteValues = [10, 20];
+            const NotesWithAZero = helpers.makeTestNotes(makerNoteValues, takerNoteValues);
+            
+            const { proofData, challenge } = bilateralProof.constructBilateralSwap(NotesWithAZero, accounts[0]);
+
+            await exceptions.catchRevert(BilateralSwap.validateBilateralSwap(proofData, challenge, t2, {
+                from: accounts[0],
+                gas: 4000000,
+            }));
+        });
+
+        it('Validate failure for using a fake challenge', async () => {
+            const { proofData } = bilateralProof.constructBilateralSwap(testNotes, accounts[0]);
+
+            const fakeChallenge = new BN(crypto.randomBytes(32), 16).umod(GROUP_MODULUS).toString(10);
+
+            await exceptions.catchRevert(BilateralSwap.validateBilateralSwap(proofData, fakeChallenge, t2, {
+                from: accounts[0],
+                gas: 4000000,
+            }));
+        });
+
+        it('Validate failure for using fake proof data', async () => {
+            const { challenge } = bilateralProof.constructBilateralSwap(testNotes, accounts[0]);
+     
+            const fakeProofData = new Array(4).map(() => new Array(6).map(() => toBytes32.randomBytes32()));
+
+            await exceptions.catchRevert(BilateralSwap.validateBilateralSwap(fakeProofData, challenge, t2, {
+                from: accounts[0],
+                gas: 4000000,
+            }));
+        });
+
+        it('Validate failure when points not on curve', async () => {
+            const zeroes = `${padLeft('0', 64)}`;
```

Comment:
We should update this to produce a satisfying proof *if* we don't check for whether points are on the curve.

i.e. the x and y coordinates for the 4 notes should be '0', but the challenge should correctly resolve to the hash of the commitments and blinding factors (also points with x/y coordinates of '0')